File: packages/jumpstarter-kubernetes/jumpstarter_kubernetes/cluster/kubectl.py
"""Kubectl operations for cluster management."""

import json
from typing import Dict, List, Optional

from ..clusters import V1Alpha1ClusterInfo, V1Alpha1ClusterList, V1Alpha1JumpstarterInstance
from .common import run_command


async def check_kubernetes_access(context: Optional[str] = None, kubectl: str = "kubectl") -> bool:
    """Check if Kubernetes cluster is accessible."""
    try:
        cmd = [kubectl]
        if context:
            cmd.extend(["--context", context])
        cmd.extend(["cluster-info", "--request-timeout=5s"])

        returncode, _, _ = await run_command(cmd)
        return returncode == 0
    except RuntimeError:
        return False


async def get_kubectl_contexts(kubectl: str = "kubectl") -> List[Dict[str, str]]:
    """Get all kubectl contexts."""
    contexts = []

    try:
        cmd = [kubectl, "config", "view", "-o", "json"]
        returncode, stdout, stderr = await run_command(cmd)

        if returncode != 0:
            from ..exceptions import KubeconfigError
            raise KubeconfigError(f"Failed to get kubectl config: {stderr}")

        config = json.loads(stdout)

        current_context = config.get("current-context", "")
        context_list = config.get("contexts", [])

        for ctx in context_list:
            context_name = ctx.get("name", "")
            cluster_name = ctx.get("context", {}).get("cluster", "")
            user_name = ctx.get("context", {}).get("user", "")
            namespace = ctx.get("context", {}).get("namespace") or "default"

            # Get cluster server URL
            server_url = ""
            for cluster in config.get("clusters", []):
                if cluster.get("name") == cluster_name:
                    server_url = cluster.get("cluster", {}).get("server", "")
                    break

            contexts.append(
                {
                    "name": context_name,
                    "cluster": cluster_name,
                    "server": server_url,
                    "user": user_name,
                    "namespace": namespace,
                    "current": context_name == current_context,
                }
            )

        return contexts

    except json.JSONDecodeError as e:
        from ..exceptions import KubeconfigError
        raise KubeconfigError(f"Failed to parse kubectl config: {e}") from e
    except Exception as e:
        from ..exceptions import KubeconfigError
        raise KubeconfigError(f"Error listing kubectl contexts: {e}") from e


async def check_jumpstarter_installation(  # noqa: C901
    context: str, namespace: Optional[str] = None, helm: str = "helm", kubectl: str = "kubectl"
) -> V1Alpha1JumpstarterInstance:
    """Check if Jumpstarter is installed in the cluster."""
    result_data = {
        "installed": False,
        "version": None,
        "namespace": None,
        "chart_name": None,
        "status": None,
        "has_crds": False,
        "error": None,
        "basedomain": None,
        "controller_endpoint": None,
        "router_endpoint": None,
    }

    try:
        # Check for Helm installation first
        helm_cmd = [helm, "list", "--all-namespaces", "-o", "json", "--kube-context", context]
        returncode, stdout, _ = await run_command(helm_cmd)

        if returncode == 0:
            # Extract JSON from output (handle case where warnings are printed before JSON)
            json_start = stdout.find("[")
            if json_start >= 0:
                json_output = stdout[json_start:]
                releases = json.loads(json_output)
            else:
                releases = json.loads(stdout)  # Fallback to original parsing
            for release in releases:
                # Look for Jumpstarter chart
                if "jumpstarter" in release.get("chart", "").lower():
                    result_data["installed"] = True
                    result_data["version"] = release.get("app_version") or release.get("chart", "").split("-")[-1]
                    result_data["namespace"] = release.get("namespace")
                    result_data["chart_name"] = release.get("name")
                    result_data["status"] = release.get("status")

                    # Try to get Helm values to extract basedomain and endpoints
                    try:
                        values_cmd = [
                            helm,
                            "get",
                            "values",
                            release.get("name"),
                            "-n",
                            release.get("namespace"),
                            "-o",
                            "json",
                            "--kube-context",
                            context,
                        ]
                        values_returncode, values_stdout, _ = await run_command(values_cmd)

                        if values_returncode == 0:
                            # Extract JSON from values output (handle warnings)
                            json_start = values_stdout.find("{")
                            if json_start >= 0:
                                json_output = values_stdout[json_start:]
                                values = json.loads(json_output)
                            else:
                                values = json.loads(values_stdout)  # Fallback

                            # Extract basedomain
                            basedomain = values.get("global", {}).get("baseDomain")
                            if basedomain:
                                result_data["basedomain"] = basedomain
                                # Construct default endpoints from basedomain
                                result_data["controller_endpoint"] = f"grpc.{basedomain}:8082"
                                result_data["router_endpoint"] = f"router.{basedomain}:8083"

                            # Check for explicit endpoints in values
                            controller_config = values.get("jumpstarter-controller", {}).get("grpc", {})
                            if controller_config.get("endpoint"):
                                result_data["controller_endpoint"] = controller_config["endpoint"]
                            if controller_config.get("routerEndpoint"):
                                result_data["router_endpoint"] = controller_config["routerEndpoint"]

                    except (json.JSONDecodeError, RuntimeError):
                        # Failed to get Helm values, but we still have basic info
                        pass

                    break

        # Check for Jumpstarter CRDs as secondary verification
        try:
            crd_cmd = [kubectl, "--context", context, "get", "crd", "-o", "json"]
            returncode, stdout, _ = await run_command(crd_cmd)

            if returncode == 0:
                # Extract JSON from CRD output (handle warnings)
                json_start = stdout.find("{")
                if json_start >= 0:
                    json_output = stdout[json_start:]
                    crds = json.loads(json_output)
                else:
                    crds = json.loads(stdout)  # Fallback
                jumpstarter_crds = []
                for item in crds.get("items", []):
                    name = item.get("metadata", {}).get("name", "")
                    if "jumpstarter.dev" in name:
                        jumpstarter_crds.append(name)

                if jumpstarter_crds:
                    result_data["has_crds"] = True
                    if not result_data["installed"]:
                        # CRDs exist but no Helm release found - manual installation?
                        result_data["installed"] = True
                        result_data["version"] = "unknown"
                        result_data["namespace"] = namespace or "unknown"
                        result_data["status"] = "manual-install"
        except RuntimeError:
            pass  # CRD check failed, continue with Helm results

    except json.JSONDecodeError as e:
        result_data["error"] = f"Failed to parse Helm output: {e}"
    except RuntimeError as e:
        result_data["error"] = f"Command failed: {e}"

    return V1Alpha1JumpstarterInstance(**result_data)


async def get_cluster_info(
    context: str,
    kubectl: str = "kubectl",
    helm: str = "helm",
    minikube: str = "minikube",
) -> V1Alpha1ClusterInfo:
    """Get comprehensive cluster information."""
    try:
        contexts = await get_kubectl_contexts(kubectl)
        context_info = None

        for ctx in contexts:
            if ctx["name"] == context:
                context_info = ctx
                break

        if not context_info:
            return V1Alpha1ClusterInfo(
                name=context,
                cluster="unknown",
                server="unknown",
                user="unknown",
                namespace="unknown",
                is_current=False,
                type="remote",
                accessible=False,
                jumpstarter=V1Alpha1JumpstarterInstance(installed=False),
                error=f"Context '{context}' not found",
            )

        # Detect cluster type
        from .detection import detect_cluster_type

        cluster_type = await detect_cluster_type(context_info["name"], context_info["server"], minikube)

        # Check if cluster is accessible
        try:
            version_cmd = [kubectl, "--context", context, "version", "-o", "json"]
            returncode, stdout, _ = await run_command(version_cmd)
            cluster_accessible = returncode == 0
            cluster_version = None

            if cluster_accessible:
                try:
                    version_info = json.loads(stdout)
                    cluster_version = version_info.get("serverVersion", {}).get("gitVersion", "unknown")
                except (json.JSONDecodeError, KeyError):
                    cluster_version = "unknown"
        except RuntimeError:
            cluster_accessible = False
            cluster_version = None

        # Check Jumpstarter installation
        if cluster_accessible:
            jumpstarter_info = await check_jumpstarter_installation(context, None, helm, kubectl)
        else:
            jumpstarter_info = V1Alpha1JumpstarterInstance(installed=False, error="Cluster not accessible")

        return V1Alpha1ClusterInfo(
            name=context_info["name"],
            cluster=context_info["cluster"],
            server=context_info["server"],
            user=context_info["user"],
            namespace=context_info["namespace"],
            is_current=context_info["current"],
            type=cluster_type,
            accessible=cluster_accessible,
            version=cluster_version,
            jumpstarter=jumpstarter_info,
        )

    except Exception as e:
        return V1Alpha1ClusterInfo(
            name=context,
            cluster="unknown",
            server="unknown",
            user="unknown",
            namespace="unknown",
            is_current=False,
            type="remote",
            accessible=False,
            jumpstarter=V1Alpha1JumpstarterInstance(installed=False),
            error=f"Failed to get cluster info: {e}",
        )


async def list_clusters(
    cluster_type_filter: str = "all",
    kubectl: str = "kubectl",
    helm: str = "helm",
    kind: str = "kind",
    minikube: str = "minikube",
) -> V1Alpha1ClusterList:
    """List all Kubernetes clusters with Jumpstarter status."""
    try:
        contexts = await get_kubectl_contexts(kubectl)
        cluster_infos = []

        for context in contexts:
            cluster_info = await get_cluster_info(context["name"], kubectl, helm, minikube)

            # Filter by type if specified
            if cluster_type_filter != "all" and cluster_info.type != cluster_type_filter:
                continue

            cluster_infos.append(cluster_info)

        return V1Alpha1ClusterList(items=cluster_infos)

    except Exception as e:
        # Return empty list with error in the first cluster
        error_cluster = V1Alpha1ClusterInfo(
            name="error",
            cluster="error",
            server="error",
            user="error",
            namespace="error",
            is_current=False,
            type="remote",
            accessible=False,
            jumpstarter=V1Alpha1JumpstarterInstance(installed=False),
            error=f"Failed to list clusters: {e}",
        )
        return V1Alpha1ClusterList(items=[error_cluster])

File: rust/jumpstarter-cluster/src/kubectl.rs
//! `kubectl`-based cluster discovery — the read-only `get cluster` / `get clusters`
//! path (`cluster/kubectl.py`): enumerate kubeconfig contexts, probe each
//! cluster's reachability + version, and detect a Jumpstarter installation.

use crate::command::run_command;
use crate::detection::detect_cluster_type;
use crate::error::{ClusterError, Result};
use crate::info::{ClusterInfo, ClusterList, JumpstarterInstance};

/// A kubeconfig context (`KubectlContext`).
#[derive(Debug, Clone)]
pub struct KubectlContext {
    pub name: String,
    pub cluster: String,
    pub server: String,
    pub user: String,
    pub namespace: String,
    pub current: bool,
}

/// Parse `kubectl config view -o json` into the context list (`get_kubectl_contexts`).
pub async fn get_kubectl_contexts(kubectl: &str) -> Result<Vec<KubectlContext>> {
    let out = run_command(&[kubectl, "config", "view", "-o", "json"]).await?;
    if !out.ok() {
        return Err(ClusterError::Kubeconfig(format!("Failed to get kubectl config: {}", out.stderr)));
    }
    let config: serde_json::Value = serde_json::from_str(&out.stdout)
        .map_err(|e| ClusterError::Kubeconfig(format!("Failed to parse kubectl config: {e}")))?;

    let current = config.get("current-context").and_then(|v| v.as_str()).unwrap_or("");
    let empty = vec![];
    let clusters = config.get("clusters").and_then(|v| v.as_array()).unwrap_or(&empty);

    let mut contexts = Vec::new();
    for ctx in config.get("contexts").and_then(|v| v.as_array()).unwrap_or(&empty) {
        let name = ctx.get("name").and_then(|v| v.as_str()).unwrap_or("").to_string();
        let inner = ctx.get("context");
        let cluster = inner.and_then(|c| c.get("cluster")).and_then(|v| v.as_str()).unwrap_or("").to_string();
        let user = inner.and_then(|c| c.get("user")).and_then(|v| v.as_str()).unwrap_or("").to_string();
        let namespace = inner
            .and_then(|c| c.get("namespace"))
            .and_then(|v| v.as_str())
            .filter(|s| !s.is_empty())
            .unwrap_or("default")
            .to_string();
        let server = clusters
            .iter()
            .find(|c| c.get("name").and_then(|v| v.as_str()) == Some(cluster.as_str()))
            .and_then(|c| c.get("cluster"))
            .and_then(|c| c.get("server"))
            .and_then(|v| v.as_str())
            .unwrap_or("")
            .to_string();
        let current = name == current;
        contexts.push(KubectlContext { name, cluster, server, user, namespace, current });
    }
    Ok(contexts)
}

/// Strip any non-JSON prefix before the first `{` (`_parse_json_with_prefix`).
fn parse_json_with_prefix(stdout: &str) -> serde_json::Result<serde_json::Value> {
    match stdout.find('{') {
        Some(i) => serde_json::from_str(&stdout[i..]),
        None => serde_json::from_str(stdout),
    }
}

/// Query Jumpstarter CR instances to confirm a full install
/// (`_check_cr_instances` + `_apply_cr_result`); fills `installed`/`namespace`/
/// `status` or `error` on `inst`.
async fn apply_cr_instances(inst: &mut JumpstarterInstance, kubectl: &str, context: &str, namespace: Option<&str>) {
    let cr = "jumpstarters.operator.jumpstarter.dev";
    match run_command(&[kubectl, "--context", context, "get", cr, "-A", "-o", "json"]).await {
        Ok(out) if out.ok() => match serde_json::from_str::<serde_json::Value>(&out.stdout) {
            Ok(data) => {
                let items = data.get("items").and_then(|v| v.as_array());
                if let Some(items) = items.filter(|a| !a.is_empty()) {
                    let cr_ns = items[0]
                        .get("metadata")
                        .and_then(|m| m.get("namespace"))
                        .and_then(|v| v.as_str())
                        .or(namespace)
                        .unwrap_or("unknown");
                    inst.installed = true;
                    inst.namespace = Some(cr_ns.to_string());
                    inst.status = Some("installed".to_string());
                }
            }
            Err(e) => inst.error = Some(format!("CR instance check failed: {e}")),
        },
        Ok(out) => {
            let detail = if out.stderr.is_empty() { &out.stdout } else { &out.stderr };
            inst.error = Some(format!("CR instance check failed (exit {}): {}", out.code, detail));
        }
        Err(e) => inst.error = Some(format!("CR instance check failed: {e}")),
    }
}

/// Detect a Jumpstarter installation via CRD scan + CR instance check
/// (`check_jumpstarter_installation`).
pub async fn check_jumpstarter_installation(
    context: &str,
    namespace: Option<&str>,
    kubectl: &str,
) -> JumpstarterInstance {
    let mut inst = JumpstarterInstance::not_installed();
    let out = match run_command(&[kubectl, "--context", context, "get", "crd", "-o", "json"]).await {
        Ok(o) => o,
        Err(e) => {
            inst.error = Some(format!("Command failed: {e}"));
            return inst;
        }
    };
    if !out.ok() {
        let detail = if out.stderr.is_empty() { &out.stdout } else { &out.stderr };
        inst.error = Some(format!("Command failed: {detail}"));
        return inst;
    }
    let crds = match parse_json_with_prefix(&out.stdout) {
        Ok(v) => v,
        Err(e) => {
            inst.error = Some(format!("Failed to parse output: {e}"));
            return inst;
        }
    };
    let empty = vec![];
    let names: Vec<String> = crds
        .get("items")
        .and_then(|v| v.as_array())
        .unwrap_or(&empty)
        .iter()
        .filter_map(|item| item.get("metadata").and_then(|m| m.get("name")).and_then(|v| v.as_str()))
        .filter(|n| n.contains("jumpstarter.dev"))
        .map(String::from)
        .collect();
    if !names.is_empty() {
        inst.has_crds = true;
        if names.iter().any(|n| n == "jumpstarters.operator.jumpstarter.dev") {
            apply_cr_instances(&mut inst, kubectl, context, namespace).await;
        }
    }
    inst
}

fn unreachable_cluster(name: &str, error: String) -> ClusterInfo {
    ClusterInfo {
        api_version: "jumpstarter.dev/v1alpha1".to_string(),
        kind: "ClusterInfo".to_string(),
        name: name.to_string(),
        cluster: "unknown".to_string(),
        server: "unknown".to_string(),
        user: "unknown".to_string(),
        namespace: "unknown".to_string(),
        is_current: false,
        r#type: "remote".to_string(),
        accessible: false,
        version: None,
        jumpstarter: JumpstarterInstance::not_installed(),
        error: Some(error),
    }
}

/// Comprehensive info for one cluster context (`get_cluster_info`).
pub async fn get_cluster_info(context: &str, kubectl: &str, minikube: &str) -> ClusterInfo {
    let contexts = match get_kubectl_contexts(kubectl).await {
        Ok(c) => c,
        Err(e) => return unreachable_cluster(context, format!("Failed to get cluster info: {e}")),
    };
    let Some(info) = contexts.into_iter().find(|c| c.name == context) else {
        return unreachable_cluster(context, format!("Context '{context}' not found"));
    };

    let cluster_type = detect_cluster_type(&info.name, &info.server, minikube).await;

    let (accessible, version) =
        match run_command(&[kubectl, "--context", context, "version", "-o", "json"]).await {
            Ok(out) if out.ok() => {
                let v = serde_json::from_str::<serde_json::Value>(&out.stdout)
                    .ok()
                    .and_then(|j| j.get("serverVersion").and_then(|s| s.get("gitVersion")).and_then(|g| g.as_str()).map(String::from))
                    .unwrap_or_else(|| "unknown".to_string());
                (true, Some(v))
            }
            _ => (false, None),
        };

    let jumpstarter = if accessible {
        check_jumpstarter_installation(context, None, kubectl).await
    } else {
        let mut inst = JumpstarterInstance::not_installed();
        inst.error = Some("Cluster not accessible".to_string());
        inst
    };

    ClusterInfo {
        api_version: "jumpstarter.dev/v1alpha1".to_string(),
        kind: "ClusterInfo".to_string(),
        name: info.name,
        cluster: info.cluster,
        server: info.server,
        user: info.user,
        namespace: info.namespace,
        is_current: info.current,
        r#type: cluster_type,
        accessible,
        version,
        jumpstarter,
        error: None,
    }
}

/// List every kubeconfig cluster with Jumpstarter status, optionally filtered by
/// type (`list_clusters`).
pub async fn list_clusters(type_filter: &str, kubectl: &str, minikube: &str) -> ClusterList {
    let contexts = match get_kubectl_contexts(kubectl).await {
        Ok(c) => c,
        Err(e) => return ClusterList::new(vec![unreachable_cluster("error", format!("Failed to list clusters: {e}"))]),
    };
    let mut items = Vec::new();
    for ctx in contexts {
        let info = get_cluster_info(&ctx.name, kubectl, minikube).await;
        if type_filter != "all" && info.r#type != type_filter {
            continue;
        }
        items.push(info);
    }
    ClusterList::new(items)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn json_prefix_is_stripped() {
        let v = parse_json_with_prefix("warn: x\n{\"a\":1}").unwrap();
        assert_eq!(v["a"], serde_json::json!(1));
        assert!(parse_json_with_prefix("{\"b\":2}").is_ok());
    }
}
